### ISSUE TITLE
"ipfs repo gc" is freezing IPFS node

### ISSUE THREAD

**Issue by obo20** [ipfs/go-ipfs#6107]
#### Version information:
go-ipfs version: 0.4.19-   (I also tried downgrading to 0.4.18 and am still running into this issue)
Repo version: 7
System version: amd64/linux
Golang version: go1.11.5
Filesystem: Flatfs

#### Type:
panic

-->

#### Description:
The most recent garbage collection I manually ran on one of our nodes completely froze the node and made it unresponsive to any sort of adding / pinning of content. 

Upon restarting the node and attempting another garbage collection, I received the same result.

It seems I'm not the only one with this issue:
https://github.com/ipfs/go-ipfs/issues/5177
https://github.com/ipfs/go-ipfs/issues/5489

**Comment by Stebalien** [comment]
Talking with @obo20 about this on IRC. Looks like something is deadlocking somewhere.

New information:

* ~15GiB of data
* ~130,000 blocks
* `ipfs repo verify` passes

That means it's not a disk issue and we can definitely iterate over the entire collection. It looks like an issue traversing the graph or maybe some locking issue.

---

Current status: The repo is live so we're cloning to a new machine where we can leave a GC running for a while (order of minutes). We should then be able to pull a goroutine list to see what's stuck.

**Comment by obo20** [comment]
A little update. After working with @Stebalien on this, once I got the machine cloned, the garbage collection on that machine ran in about 2 minutes. With this knowledge, I attempted another garbage collection on the main machine and it also ran in about 2 minutes.

We're still not sure what caused the original garbage collection non-ending deadlock, but the issue seems to have resolved itself for now.

I did learn something through this process though, in that IPFS intentionally locks the node down for things like adding files when doing a GC. I'd love to see this paradigm changed, as this could be pretty problematic at scale, especially for production nodes. I'm going to file a new issue on this so things don't get too mixed up.
